bindings/c/test/unit/unit_tests.cpp

```diff
@@ -0,0 +1,1840 @@
+//
+//   fdb::Transaction tr(db);
+//   tr.set((const uint8_t *)"foo", 3, (const uint8_t *)"bar", 3);
+//   fdb::EmptyFuture f1 = tr.commit();
+//
+//   CHECK(wait_future(f1) == 2006); // invalid_option_value
+//
+//   // Set size limit back to default.
+//   size_limit = 10000000;
+//   fdb_check(fdb_database_set_option(db, FDB_DB_OPTION_TRANSACTION_SIZE_LIMIT,
+//                                     (const uint8_t *)&size_limit,
+//                                     sizeof(size_limit)));
+// }
+
+TEST_CASE("FDB_DB_OPTION_TRANSACTION_SIZE_LIMIT") {
+  int64_t size_limit = 32;
+  fdb_check(fdb_database_set_option(db, FDB_DB_OPTION_TRANSACTION_SIZE_LIMIT,
+                                    (const uint8_t *)&size_limit,
+                                    sizeof(size_limit)));
+
+  fdb::Transaction tr(db);
+  tr.set("foo", "foundation database is amazing");
+  fdb::EmptyFuture f1 = tr.commit();
+
+  CHECK(wait_future(f1) == 2101); // transaction_too_large
+
+  // Set size limit back to default.
+  size_limit = 10000000;
+  fdb_check(fdb_database_set_option(db, FDB_DB_OPTION_TRANSACTION_SIZE_LIMIT,
+                                    (const uint8_t *)&size_limit,
+                                    sizeof(size_limit)));
+}
+
+TEST_CASE("fdb_transaction_set_read_version old_version") {
+  fdb::Transaction tr(db);
+
+  tr.set_read_version(1);
+  fdb::ValueFuture f1 = tr.get("foo", /*snapshot*/ true);
+
+  fdb_error_t err = wait_future(f1);
+  CHECK(err == 1007); // transaction_too_old
+}
+
+TEST_CASE("fdb_transaction_set_read_version future_version") {
+  fdb::Transaction tr(db);
+
+  tr.set_read_version(1UL << 62);
+  fdb::ValueFuture f1 = tr.get("foo", /*snapshot*/ true);
+
+  fdb_error_t err = wait_future(f1);
+  CHECK(err == 1009); // future_version
+}
+
+TEST_CASE("fdb_transaction_get_range reverse") {
+  std::map<std::string, std::string> data =
+      create_data({ { "a", "1" }, { "b", "2" }, { "c", "3" }, { "d", "4" } });
+  insert_data(db, data);
+
+  fdb::Transaction tr(db);
+  while (1) {
+    auto result = get_range(
+        tr, FDB_KEYSEL_FIRST_GREATER_OR_EQUAL(
+          (const uint8_t*)KEY("a").c_str(),
+          KEYSIZE("a")
+        ),
+        FDB_KEYSEL_LAST_LESS_OR_EQUAL(
+          (const uint8_t*)KEY("d").c_str(),
+          KEYSIZE("d")
+        ) + 1, /* limit */ 0, /* target_bytes */ 0,
+        /* FDBStreamingMode */ FDB_STREAMING_MODE_WANT_ALL, /* iteration */ 0,
+        /* snapshot */ false, /* reverse */ 1);
+
+    if (result.err) {
+      fdb::EmptyFuture f1 = tr.on_error(result.err);
+      fdb_check(wait_future(f1));
+      continue;
+    }
+
+    CHECK(result.kvs.size() > 0);
+    CHECK(result.kvs.size() <= 4);
+    if (result.kvs.size() < 4) {
+      CHECK(result.more);
+    }
+
+    // Read data in reverse order, keeping in mind that out_count might be
+    // smaller than requested.
+    auto it = data.rbegin();
+    std::advance(it, data.size() - result.kvs.size());
```

Comment:
Why do we want to advance the iterator here?

Reply:
In case less data is returned than expected, you need to advance the reverse iterator to start at the correct point.

Reply:
I was wrong, even if less data is returned than expected, the results still start at the end of the range.